File: haskell/triangle/src/Triangle.hs
module Triangle (TriangleType(..), triangleType) where

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Illegal
                  deriving (Eq, Show)

triangleType :: (Num a, Ord a) => a -> a -> a -> TriangleType
triangleType a b c 
    | or [a <= 0, b <=0, c <= 0] = Illegal
    | or [a + b <= c, b + c <= a, a + c <= b] = Illegal
    | otherwise =
        if all id pairs then 
            Equilateral else
            if any id pairs then 
                Isosceles
            else
                Scalene
            where
                pairs = [a == b, a == c, b == c]

File: ruby/triangle/triangle.rb
class Triangle

  attr_reader :pairsEquality, :valid

  def initialize(ary)
    @valid = ary.all? {|x| x != 0} && ary.permutation(3).all?{|a| a[0] + a[1] >= a[2]}
    @pairsEquality = ary.combination(2).to_a.map {|x| x[0] == x[1]}
  end

  def equilateral?
    valid && pairsEquality.all? {|x| x}
  end

  def isosceles?
    valid && pairsEquality.any? {|x| x}
  end

  def scalene?
    valid && !isosceles?
  end
end
